### PR TITLE
Update visibilitychange data to reflect Safari 14 fix

### PR DESCRIPTION
**Pull request by sideshowbarker**
https://trac.webkit.org/changeset/267614/webkit shipped in Safari 14 and fixed as problem in Safari’s `visibilitychange` behavior which had led to it being marked as `partial_implementation` in BCD, with an associated Note.

So this change updates BCD to reflect that as of version 14, Safari has full support for `visibilitychange` and `onvisibilitychange`.

### CODE REVIEW

File: api/Document.json [hunk 1]
```diff
@@ -11305,24 +11315,34 @@
...
+            "safari": [
+              {
+                "version_added": "14"
+              },
+              {
+                "version_added": "7",
+                "partial_implementation": true,
+                "notes": [
+                  "Doesn't fire the <code>visibilitychange</code> event when navigating away from a document, so also include code to check for the <code>pagehide</code> event (which does fire for that case in all current browsers). See WebKit bugs <a href='https://webkit.org/b/116769'>116769</a>, <a href='https://webkit.org/b/151234'>151234</a>, <a href='https://webkit.org/b/151610'>151610</a>, and <a href='https://webkit.org/b/194897'>194897</a>.",
+                  "Before Safari 14, the event does not bubble, so <code>document.addEventListener('visibilitychange', ...)</code> works, but <code>window.addEventListener('visibilitychange', ...)</code> does not.",
```
Comment (foolip) [inline api/Document.json:11327]: The Safari 14 part is implied now and can be dropped.

**Review by foolip — COMMENTED** [review]
Looks good! Can you share how you got the Safari version from the revision?

File: api/Document.json [hunk 2]
```diff
@@ -7930,22 +7930,32 @@
...
-              "version_added": "10.3",
-              "partial_implementation": true,
-              "notes": [
-                "Doesn't fire the <code>visibilitychange</code> event when navigating away from a document, so also include code to check for the <code>pagehide</code> event (which does fire for that case in all current browsers). See WebKit bugs <a href='https://webkit.org/b/116769'>116769</a>, <a href='https://webkit.org/b/151234'>151234</a>, <a href='https://webkit.org/b/151610'>151610</a>, and <a href='https://webkit.org/b/194897'>194897</a>.",
-                "Before Safari iOS 10.3, this event handler attribute was not supported; however, the event itself was supported since Safari iOS 7. The event can be listened to via <code>document.addEventListener('visibilitychange', function() {});</code>."
-              ]
-            },
+            "safari": [
+              {
+                "version_added": "14"
```
Comment (foolip) [inline api/Document.json:7935]:
```suggestion
                "version_added": "14.1"
```

File: api/Document.json [hunk 3]
```diff
@@ -7930,22 +7930,32 @@
...
+                "partial_implementation": true,
+                "notes": [
+                  "Doesn't fire the <code>visibilitychange</code> event when navigating away from a document, so also include code to check for the <code>pagehide</code> event (which does fire for that case in all current browsers). See WebKit bugs <a href='https://webkit.org/b/116769'>116769</a>, <a href='https://webkit.org/b/151234'>151234</a>, <a href='https://webkit.org/b/151610'>151610</a>, and <a href='https://webkit.org/b/194897'>194897</a>.",
+                  "Before Safari 10.1, this event handler attribute was not supported; however, the event itself was supported since Safari 7. The event can be listened to via <code>document.addEventListener('visibilitychange', function() {});</code>."
+                ]
+              }
+            ],
+            "safari_ios": [
+              {
+                "version_added": "14"
```
Comment (foolip) [inline api/Document.json:7948]:
```suggestion
                "version_added": "14.5"
```

File: api/Document.json [hunk 4]
```diff
@@ -11305,24 +11315,34 @@
...
-              "partial_implementation": true,
-              "notes": [
-                "Doesn't fire the <code>visibilitychange</code> event when navigating away from a document, so also include code to check for the <code>pagehide</code> event (which does fire for that case in all current browsers). See WebKit bugs <a href='https://webkit.org/b/116769'>116769</a>, <a href='https://webkit.org/b/151234'>151234</a>, <a href='https://webkit.org/b/151610'>151610</a>, and <a href='https://webkit.org/b/194897'>194897</a>.",
-                "Before Safari iOS 14, the event does not bubble, so <code>document.addEventListener('visibilitychange', ...)</code> works, but <code>window.addEventListener('visibilitychange', ...)</code> does not.",
-                "Before Safari iOS 10.3, the <code>onvisibilitychange</code> attribute is not supported. To listen to this event in earlier versions of Safari iOS, use <code>document.addEventListener('visibilitychange', function() {});</code>."
-              ]
-            },
+            "safari": [
+              {
+                "version_added": "14"
```
Comment (foolip) [inline api/Document.json:11320]:
```suggestion
                "version_added": "14.1"
```

File: api/Document.json [hunk 5]
```diff
@@ -11305,24 +11315,34 @@
...
+                "notes": [
+                  "Doesn't fire the <code>visibilitychange</code> event when navigating away from a document, so also include code to check for the <code>pagehide</code> event (which does fire for that case in all current browsers). See WebKit bugs <a href='https://webkit.org/b/116769'>116769</a>, <a href='https://webkit.org/b/151234'>151234</a>, <a href='https://webkit.org/b/151610'>151610</a>, and <a href='https://webkit.org/b/194897'>194897</a>.",
+                  "Before Safari 14, the event does not bubble, so <code>document.addEventListener('visibilitychange', ...)</code> works, but <code>window.addEventListener('visibilitychange', ...)</code> does not.",
+                  "Before Safari 10.1, the <code>onvisibilitychange</code> attribute is not supported. To listen to this event in earlier versions of Safari, use <code>document.addEventListener('visibilitychange', function() {});</code>."
+                ]
+              }
+            ],
+            "safari_ios": [
+              {
+                "version_added": "14"
```
Comment (foolip) [inline api/Document.json:11334]:
```suggestion
                "version_added": "14.5"
```

**Review by foolip — CHANGES_REQUESTED** [review]
Trying to confirm the version I get trunk version 611.1.1 which should map to 14.1/14.5:
https://trac.webkit.org/changeset/267614/webkit
https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/Configurations/Version.xcconfig?rev=267614

There's also an unresolved comment about a note.

**Review by foolip — DISMISSED** [review]

File: api/Document.json (hunk 1 above)
Comment (foolip) [reply api/Document.json:11327]:
```suggestion
                  "The event does not bubble, so <code>document.addEventListener('visibilitychange', ...)</code> works, but <code>window.addEventListener('visibilitychange', ...)</code> does not.",
```

**Comment by foolip** [comment]
Oh, one more thing. Can you add `version_removed` to the partial entry and take a look over the notes?

**Comment by sideshowbarker** [comment]
> Hey @sideshowbarker, do you plan on returning to this PR?

I do — maybe later this week, maybe early next week

**Comment by ddbeck** [comment]
This seems _this_ close to being mergeable. As I see it, two things need to happen:

- [ ] The proposed note change needs to be adopted
- [ ] The partial statements need to have `version_removed` set to the `version_added` of the non-partial statements

@sideshowbarker do you want to do these things? Or would you like someone to take this over?

**Comment by sideshowbarker** [comment]
> This seems _this_ close to being mergeable. As I see it, two things need to happen:
> 
>     * [ ]  The proposed note change needs to be adopted
> 
>     * [ ]  The partial statements need to have `version_removed` set to the `version_added` of the non-partial statements
> 
> 
> @sideshowbarker do you want to do these things? Or would you like someone to take this over?

I’ve gone ahead and pushed an update here with those changes.

**Review by ddbeck — APPROVED** [review]
Thank you! 🎉

**Pull request merged**